src/Sylius/Component/Core/OrderProcessing/StateResolver.php

```diff
@@ -15,46 +15,53 @@
 use Sylius\Component\Core\Model\OrderShippingStates;
 use Sylius\Component\Core\Model\PaymentInterface;
 use Sylius\Component\Core\Model\ShipmentInterface;
+use SM\Factory\Factory;
+use Sylius\Component\Order\OrderTransitions;
+use Sylius\Component\Core\OrderCheckoutStates;
 
 /**
  * @author Paweł Jędrzejewski <pawel@sylius.org>
  */
 class StateResolver implements StateResolverInterface
 {
+    private $smFactory;
+
+    public function __construct(Factory $smFactory)
+    {
+        $this->smFactory = $smFactory;
+    }
+
     /**
      * {@inheritdoc}
      */
     public function resolvePaymentState(OrderInterface $order)
     {
-        $paymentState = PaymentInterface::STATE_NEW;
+        // do not recalculate payment state. althoguh the payment may have been
+        // refunded, the payment WAS completed for this order.
+        if ($order->getPaymentState() === OrderInterface::STATE_PAYMENT_COMPLETED) {
+            return;
+        }
+
+        // do not calculate payment state when checkout has not been completed
+        if ($order->getCheckoutState() === OrderCheckoutStates::STATE_COMPLETED) {
+            return;
+        }
```

Comment:
The above two cases assume the COMPLETED is the final *payment state* of the order, which similifies things here, but is probably insufficent.

I guess we should take into account refunded, failed etc also.

```
if (all payments are refunded) state = REFUNDED
if (some payments are refunded) state = PARTIALLY_REFUNDED
if (some payments failed) state = PENDING
if (all paymens failed) state = FAILED
```

?